### PR TITLE
[MPS] Disallow reshape in slice

### LINKED ISSUE #95883: Failed assertion `[MPSNDArrayDescriptor sliceDimension:withSubrange:] error: the range subRange.start + subRange.length does not fit in dimension[2] (2)' on gradient computation through Linear with transpose on MPS
### 🐛 Describe the bug

The following code, when run using MPS, produces the following internal MPS error:
```
/AppleInternal/Library/BuildRoots/a0876c02-1788-11ed-b9c4-96898e02b808/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:88: 
failed assertion `[MPSNDArrayDescriptor sliceDimension:withSubrange:] error: the range subRange.start + subRange.length does not fit in dimension[2] (2)'
```
The error does not occur when using the CPU device, nor when using the second commented-out `lin()` invocation with the explicit `.reshape()`s instead of the first, nor when removing the transpose or concatenation operations.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

B = 4
T = 1
device = "mps"

lin = nn.Linear(10, 256).to(device)
x = torch.rand([B, T, 10], device=device, requires_grad=True)
x = lin(x)
# x = lin(x.reshape(B * T, -1)).reshape(B, T, -1)  # does not error

assert x.shape == (B, T, 256)

cls_token = torch.rand([1, 256], device=device, requires_grad=True).repeat(B, 1, 1)
x = torch.cat([cls_token, x], dim=1)

x = x.transpose(0, 1) # if this line is commented out, does not error either

loss = F.mse_loss(x, torch.rand_like(x))

loss.backward()
```

### Versions

```
PyTorch version: 2.1.0.dev20230302
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.102)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:38:11)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-12.6-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==1.9.3
[pip3] torch==2.1.0.dev20230302
[pip3] torchmetrics==0.11.3
[pip3] torchvision==0.15.0.dev20230302
[conda] numpy                     1.24.2           py39hff61c6a_0    conda-forge
[conda] pytorch                   2.1.0.dev20230302         py3.9_0    pytorch-nightly
[conda] pytorch-lightning         1.9.3                    pypi_0    pypi
[conda] torchmetrics              0.11.3                   pypi_0    pypi
[conda] torchvision               0.15.0.dev20230302        py39_cpu    pytorch-nightly
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by DenisVieriu97**
Disallow reshapes for arrayViews.
Current code allows a base shape of `[2, 4, 256]` to be sliced into `[4, 1, 256]` (view's shape) - which is not possible. Slicing a smaller dimension into a bigger one will always error out.  

Fixes https://github.com/pytorch/pytorch/issues/95883

### CODE REVIEW

**Review by razarmehr — APPROVED** [review]
LGTM

**Review by kulinseth — APPROVED** [review]
Please add the other 2 tests.
